idioms/coercion-arguments.md

```diff
@@ -0,0 +1,115 @@
+# Use borrowed types for arguments
+
+## Description
+
+Using a target of a deref coercion can increase the flexibility of your code when you are deciding which argument type to use for a function argument.
+In this way, the function will accept more input types.
+
+For example; using `&str` instead of a `&String`, or `&[T]` in preference of `&Vec<T>`.  This is not limited to slice-able or fat pointer types. In fact you should always prefer using the __borrowed type__ over __borrowing the owned type__, e.g., `&T` to `&Box<T>`.
+
+Using borrowed types you can avoid layers of indirection for those instances where the owned type already provides a layer of indirection, as it will be illustrated in the following examples. For instance, a `String` has a layer of indirection, so a `&String` will have two layers of indrection.
+We can avoid this by using `&str` instead, and letting `&String` coerce to a `&str` whenever the function is invoked. 
+
+## Example
+
+For this example, we will illustrate some differences for using `&String` as a function argument versus using a `&str`, but the ideas apply as well to using `&Vec<T>` versus using a `&[T]` or using a `&T` versus a `&Box<T>`.
+
+Consider an example where we wish to determine if a word contains three consecutive vowels.
+We don't need to own the string to determine this so we will take a reference.
+
+The code might look something like this:
+
+```rust
+fn three_vowels(word: &String) -> bool {
+    let mut vowel_count = 0;
+    for c in word.chars() {
+        match c {
+            'a' | 'e' | 'i' | 'o' | 'u' => {
+                vowel_count += 1;
+                if vowel_count >= 3 {
+                    return true
+                }
+            }
+            _ => vowel_count = 0
+        }
+    }
+    false
+}
+
+fn main() {
+    let ferris = "Ferris".to_string();
+    let curious = "Curious".to_string();
+    println!("{}: {}", ferris, three_vowels(&ferris));
+    println!("{}: {}", curious, three_vowels(&curious));
+
+    // Using a `&String` type in our arguent we will find the following example fails:
+    // println!("Ferris: {}", three_vowels("Ferris"));
+    // println!("Curious: {}", three_vowels("Curious"));
+
+}
+```
```

Comment:
```suggestion
}
TODO REMOVE ME```

This works fine because we are passing a `&String` type as a parameter.

What if we try to pass `&str` instead?

TODO REMOVE ME```rust,ignore
println!("Ferris: {}", three_vowels("Ferris"));
println!("Curious: {}", three_vowels("Curious"));
TODO REMOVE ME```
```

Reply:
I had to insert TODO REMOVE ME because if you want to do github suggestions you can't mess up with code blocks

Reply:
I've put that inside the function before and commented it out to let people try it in the playground.

Reply:
Pushed the last version I wrote now, can you read it if it's fine?